### PR TITLE
feat(#1507) add guide for testing dbt models 

### LINKED ISSUE #1507: Document DBT model testing in CHT Pipeline
- Introduction to DBT Testing
- Types of tests in DBT Core
- Test implementation examples
- Test execution

Related to [cht-pipeline #150](https://github.com/medic/cht-pipeline/issues/150#top)

### PR DESCRIPTION
**Pull request by lorerod**
# Description

Closes #1507 

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by njuguna-n** [comment]
@lorerod I am out of office tomorrow so I'll review this on Monday if that's okay.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+- [Singular tests](https://docs.getdbt.com/docs/build/data-tests#singular-data-tests): These are written in a SQL file with a query that returns records that fail the test. This type of test is straightforward and focuses on specific conditions or rules that data must meet.
+
+### Unit Tests
+
+Unit tests are essential for validating complex SQL logic and transformations in dbt models. These tests are especially valuable when working with intricate SQL expressions. They help catch errors before deploying changes, ensuring the logic behaves as expected, particularly in critical models or when handling edge cases.
+
+For more details on formatting unit tests, refer to the [official dbt documentation](https://docs.getdbt.com/reference/resource-properties/unit-tests).
+
+
+## Test Recommendations for the CHT Pipeline
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:34]:
```suggestion
## Guidelines for CHT Pipeline tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+Adding dbt tests to your workflows will check code accuracy and data integrity. It aims to prevent data quality regressions when dbt models are written or modified. And also very important, it will help the team stay engaged with the data and adapt more quickly. By including testing during the development process, teams ensure their data changes are solid.
+
+## Types of tests in DBT Core
+
+In DBT Core, there are two main types of tests: [data tests](https://docs.getdbt.com/docs/build/data-tests) and [unit tests](https://docs.getdbt.com/docs/build/unit-tests). The main difference is that data tests are meant to be executed with every pipeline run to validate data integrity, and unit tests are meant to be executed with every CI run to validate transformation logic integrity.
+
+Data tests ensure your warehouse data meets specific criteria and are run at every data refresh; unit tests allow you to validate your SQL modeling logic on a small set of static inputs (typically defined using seeds or fixtures) before you materialize your full model in production.
+
+### Data Tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:19]:
```suggestion
### Data tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+Data tests can be further divided into two types:
+- Generic tests: These are foundational tests provided by DBT core, focusing on basic schema validation and source freshness. DBT core provides [four built-in generic tests](https://docs.getdbt.com/docs/build/data-tests#generic-data-tests) that are essential for data modeling and ensuring data integrity. Generic tests can accept [additional test configurations](https://docs.getdbt.com/reference/data-test-configs).
+It is also possible to define your own [custom generic tests](https://docs.getdbt.com/best-practices/writing-custom-generic-tests).
+
+- [Singular tests](https://docs.getdbt.com/docs/build/data-tests#singular-data-tests): These are written in a SQL file with a query that returns records that fail the test. This type of test is straightforward and focuses on specific conditions or rules that data must meet.
+
+### Unit Tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:27]:
```suggestion
### Unit tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+Unit tests are essential for validating complex SQL logic and transformations in dbt models. These tests are especially valuable when working with intricate SQL expressions. They help catch errors before deploying changes, ensuring the logic behaves as expected, particularly in critical models or when handling edge cases.
+
+For more details on formatting unit tests, refer to the [official dbt documentation](https://docs.getdbt.com/reference/resource-properties/unit-tests).
+
+
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:33]:
```suggestion
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+## Test Recommendations for the CHT Pipeline
+To ensure data integrity and the reliability of our dbt models in the CHT Pipeline, it is essential to follow these testing guidelines:
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:35]:
```suggestion
To ensure data integrity and the reliability of the dbt models in the CHT Pipeline, it is essential to follow these testing guidelines:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+- Basic generic tests for all models:
+Every model should have generic tests to enforce critical constraints and relationships. Use dbt core build in generic tests.
+
+- Singular or custom generic data tests for aggregation models:
+For models that perform data aggregation, it is crucial to include singular data tests or custom generic data tests to ensure that the aggregated data meets the expected criteria. These tests help verify that:
+  - Data is accurately aggregated according to the defined logic.
+  - The results align with business expectations and requirements.
+  - Custom data tests are particularly important for aggregation models, as errors in these models can lead to significant discrepancies in reports and analyses.
+
+- Unit Tests for complex logic:
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:46]:
```suggestion
- Unit tests for complex logic:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+Unit tests are not strictly required but are highly recommended, especially for models with complex transformation logic. Examples of when to use unit tests include:
+  - Complex SQL Logic: Such as regex, date math, window functions, or extensive CASE WHEN statements.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:48]:
```suggestion
  - Complex SQL Logic: Such as regex, date math, window functions, or extensive `CASE WHEN` statements.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  - Custom Calculations: When creating functions or applying unique data processing logic.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:49]:
```suggestion
  - Custom calculations: When creating functions or applying unique data processing logic.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  - Edge Cases: To handle scenarios that are not typically found in actual data but may arise unexpectedly.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:50]:
```suggestion
  - Edge cases: To handle scenarios that are not typically found in actual data but may arise unexpectedly.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+## Test implementation in CHT pipeline
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:52]:
```suggestion
## Writing CHT Pipeline tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+Have schema tests for all models and specifically enforcing unique, not_null, and relationships where relevant. (All models should have these tests)
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:54]:
```suggestion

- All models should have **schema tests**, specifically enforcing `unique`, `not_null`, and `relationships` where relevant.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+Add custom data tests for aggregation models to ensure the data is as expected (Only specific models need these tests)
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:55]:
```suggestion

- Custom data tests for aggregation models should be added to ensure the data is as expected. Only specific models need these tests.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+Add unit tests for complex logic e.g. data math, window functions e.t.c. (not required but good to have especially if model is covered by data tests)
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:56]:
```suggestion

- Add unit tests for complex logic, such as data math or window functions. These are not required but are good to have, especially if the model is covered by data tests. 

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+Our project is organized with a `/models` directory containing SQL files and YAML files for generic tests, and a `/test` directory with folders for fixtures and singular tests.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:58]:
```suggestion
[CHT Pipeline](https://github.com/medic/cht-pipeline) contains a `/models` directory containing SQL files and YAML files for generic tests and a `/test` directory with folders for fixtures and singular tests.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            contacts.yml      --> generic data tests
+            patient.sql
+            person.sql
+            place.sql
+        /forms
+        /meta
+        /root
+        /users
+    /test
+        /fixtures                                     --> input data for unit tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:77]:
```suggestion
        /fixtures             --> input data for unit tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+            data_record/
+                data_record_initial_expected.csv
+                data_record_source_table_initial.csv
+                data_record_document_metadata_initial.csv
+        /sqltest                                       --> singular data tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:82]:
```suggestion
        /sqltest              --> singular data tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+            contact.sql
+            data_record.sql
+            patient.sql
+```
+
+### Data Tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:88]:
```suggestion
### Writing data tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+***Generic data tests*** are included in the model's YAML file under the `data-tests` tag, ensuring basic validations like `relationships`, `not_null`, and `unique`. The following image contains an extract of the properties of the  `contacts.yml` file.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:90]:
```suggestion
**Generic data tests** are included in the model's YAML file under the `data-tests` tag, ensuring basic validations like `relationships`, `not_null`, and `unique`. The following code block contains an extract of the properties of the  `contacts.yml` file.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+{{< figure src="contacts-yml.png" link="contacts-yml.png" class=" center col-16 col-lg-12" >}}
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:91]: Please replace the image with a code block. This will increase readability, make the page consistent with the rest of the docs site, and support version control changes easier.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+
+***Singular data tests*** are located in the `/test/sqltest/` folder and check data synchronization between source and final tables. For example, the following image shows the test to ensure that data between the `source` table and `contact` tables is synchronized correctly.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:93]:
```suggestion
**Singular data tests** are located in the `/test/sqltest/` folder and check data synchronization between source and final tables. The code block below contains a test to ensure that data between the `source` table and `contact` tables is synchronized correctly.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+Records that should be deleted are properly removed from both tables.
+All relevant fields between the two tables match, preserving data accuracy and integrity.
+{{< figure src="contact-sql.png" link="contact-sql.png" class=" center col-16 col-lg-12" >}}
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:96]: Please replace the image with a code block. This will increase readability, make the page consistent with the rest of the docs site, and support version control changes easier.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+### Unit Tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:98]:
```suggestion
### Writing unit tests
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+
+Unit tests are defined in a YML file inside a `tests` folder in each `model` group folder. The input format of choice is CSV using fixtures, defined in `/test/fixture` folder.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:100]:
```suggestion
Unit tests are defined in a YAML file inside a `tests` folder in each `model` group folder. The input format of choice is CSV using fixtures, defined in `/test/fixture` folder.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+
+The following image shows the content in `/models/contacts/tests/contacts.yml` file showing the test to validate transformation and data integrity for `contact` model.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:102]:
```suggestion
The code block below is an extract from the `/models/contacts/tests/contacts.yml` file containing the test to validate transformation and data integrity for the `contact` model.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+{{< figure src="contact-unit-test.png" link="contact-unit-test.png" class=" center col-16 col-lg-12" >}}
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:103]: Please replace the image with a code block. This will increase readability, make the page consistent with the rest of the docs site, and support version control changes easier.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+
+The following images shows the content of the input fixtures.
+{{< figure src="contact-document-metadata-initial.png" link="contact-document-metadata-initial.png" class=" center col-16 col-lg-12" >}}{{< figure src="contact-source-table-initial.png" link="contact-source-table-initial.png" class=" center col-16 col-lg-12" >}}
+
+And finally, the fixture representing the expected data:
+{{< figure src="contact-initial-expected.png" link="contact-initial-expected.png" class=" center col-16 col-lg-12" >}}
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:109]: Please replace the image with a code block. This will increase readability, make the page consistent with the rest of the docs site, and support version control changes easier.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+### Unit Tests
+
+Unit tests are defined in a YML file inside a `tests` folder in each `model` group folder. The input format of choice is CSV using fixtures, defined in `/test/fixture` folder.
+
+The following image shows the content in `/models/contacts/tests/contacts.yml` file showing the test to validate transformation and data integrity for `contact` model.
+{{< figure src="contact-unit-test.png" link="contact-unit-test.png" class=" center col-16 col-lg-12" >}}
+
+The following images shows the content of the input fixtures.
+{{< figure src="contact-document-metadata-initial.png" link="contact-document-metadata-initial.png" class=" center col-16 col-lg-12" >}}{{< figure src="contact-source-table-initial.png" link="contact-source-table-initial.png" class=" center col-16 col-lg-12" >}}
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:106]: Please replace the image with a code block. This will increase readability, make the page consistent with the rest of the docs site, and support version control changes easier.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+
+And finally, the fixture representing the expected data:
+{{< figure src="contact-initial-expected.png" link="contact-initial-expected.png" class=" center col-16 col-lg-12" >}}
+
+## Test execution using Docker
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:111]:
```suggestion
## Running tests with Docker
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+
+Unit tests are defined in a YML file inside a `tests` folder in each `model` group folder. The input format of choice is CSV using fixtures, defined in `/test/fixture` folder.
+
+The following image shows the content in `/models/contacts/tests/contacts.yml` file showing the test to validate transformation and data integrity for `contact` model.
+{{< figure src="contact-unit-test.png" link="contact-unit-test.png" class=" center col-16 col-lg-12" >}}
+
+The following images shows the content of the input fixtures.
+{{< figure src="contact-document-metadata-initial.png" link="contact-document-metadata-initial.png" class=" center col-16 col-lg-12" >}}{{< figure src="contact-source-table-initial.png" link="contact-source-table-initial.png" class=" center col-16 col-lg-12" >}}
+
+And finally, the fixture representing the expected data:
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:108]:
```suggestion
The fixture below represents the expected data:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+{{< figure src="contact-initial-expected.png" link="contact-initial-expected.png" class=" center col-16 col-lg-12" >}}
+
+## Test execution using Docker
+
+By running dbt tests in a Dockerized environment, we isolate dependencies and configurations, making the testing process more reproducible and easier to manage. These scripts and configuration files work together to automate the process of setting up a testing environment, running dbt models and tests:
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:113]:
```suggestion
dbt tests are run with Docker, to isolate dependencies and configurations, making the testing process reproducible and easier to manage. The scripts and configuration files below work together to automate the process of setting up a testing environment, and running dbt models and tests:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+
+`docker-compose.yml`: Defines Docker services. It starts by running PostgreSQL with the initial setup using init.sql and then builds and runs a dbt Docker container configured to connect to PostgreSQL.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:115]:
```suggestion

- `docker-compose.yml`: Defines Docker services. It starts by running PostgreSQL with the initial setup using `init.sql` and then builds and runs a dbt Docker container configured to connect to PostgreSQL.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+
+`init.sql`: A SQL script that initializes the PostgreSQL database by creating the required schema and table.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:117]:
```suggestion

- `init.sql`: A SQL script that initializes the PostgreSQL database by creating the required schema and table.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+
+`dbt/Dockerfile`: Builds the dbt Docker image. It sets up the Python environment, installs dbt and its dependencies, and copies the necessary project files into the container.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:119]:
```suggestion

- `dbt/Dockerfile`: Builds the dbt Docker image. It sets up the Python environment, installs dbt and its dependencies, and copies the necessary project files into the container.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+
+`run_dbt_tests.sh`: This script orchestrates the process of running dbt tests using Docker. It starts PostgreSQL, waits for it to initialize, runs the dbt container to execute tests, and then cleans up by stopping and removing all containers.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:121]:
```suggestion

- `run_dbt_tests.sh`: Orchestrates the process of running dbt tests using Docker. It starts PostgreSQL, waits for it to initialize, runs the dbt container to execute tests, and cleans up by stopping and removing all containers.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+
+`run_dbt_tests_docker.sh`: This script runs inside the dbt container. It sets the profiles directory, installs dbt dependencies, runs the dbt models, and ***executes the tests***.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:123]:
```suggestion

- `run_dbt_tests_docker.sh`: Runs inside the dbt container. It sets the profiles directory, installs dbt dependencies, runs the dbt models, and executes the tests.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 36]
```diff
@@ -0,0 +1,138 @@
...
+
+`profile.yml`: The dbt profile configuration file that defines the database connection settings, using environment variables to connect to the PostgreSQL database.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:125]:
```suggestion

- `profile.yml`: The dbt profile configuration file that defines the database connection settings, using environment variables to connect to the PostgreSQL database.

```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 37]
```diff
@@ -0,0 +1,138 @@
...
+
+### Prerequisites
+- `Docker`
+
+### Run the tests
+
+1. Navigate to `test` folder.
+2. Run the test script
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:133]:
```suggestion
2. Run the test script:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 38]
```diff
@@ -0,0 +1,138 @@
...
+`run_dbt_tests.sh`: This script orchestrates the process of running dbt tests using Docker. It starts PostgreSQL, waits for it to initialize, runs the dbt container to execute tests, and then cleans up by stopping and removing all containers.
+
+`run_dbt_tests_docker.sh`: This script runs inside the dbt container. It sets the profiles directory, installs dbt dependencies, runs the dbt models, and ***executes the tests***.
+
+`profile.yml`: The dbt profile configuration file that defines the database connection settings, using environment variables to connect to the PostgreSQL database.
+
+### Prerequisites
+- `Docker`
+
+### Run the tests
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:130]: Can we add a short sentence at the end that explains quickly what the developer should see after running the scripts? Something like: Docker container starts, tests run, tests pass, Docker containers are stopped and removed.

**Review by andrablaj — CHANGES_REQUESTED** [review]
This is an excellent start of this documentation @lorerod, thanks for putting it together. I added comments mostly about language consistency and format of the page. I also recommend replacing images with actual code blocks for better version control and readability.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 39]
```diff
@@ -0,0 +1,306 @@
...
+There are two main types of dbt tests:
+
+- [**data tests**](https://docs.getdbt.com/docs/build/data-tests) - meant to be executed with every pipeline run to validate data integrity. They ensure your warehouse data meets specific criteria and are run at every data refresh.
+- [**unit tests**](https://docs.getdbt.com/docs/build/unit-tests) - meant to be executed with every CI run to validate transformation logic integrity. They allow you to validate your SQL modeling logic on a small set of static inputs (typically defined using seeds or fixtures) before you materialize your full model in production.
+
+
+### Data tests
+
+Data tests can be further divided into two types:
+- [**generic tests**](https://docs.getdbt.com/docs/build/data-tests#generic-data-tests): These are foundational tests provided by dbt core, focusing on basic schema validation and source freshness. dbt core provides four built-in generic tests that are essential for data modeling and ensuring data integrity. Generic tests can accept [additional test configurations](https://docs.getdbt.com/reference/data-test-configs).
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:24]:
```suggestion
- [**generic tests**](https://docs.getdbt.com/docs/build/data-tests#generic-data-tests): These are foundational tests provided by dbt core, focusing on basic schema validation and source freshness. Dbt core provides four built-in generic tests that are essential for data modeling and ensuring data integrity. Generic tests can accept [additional test configurations](https://docs.getdbt.com/reference/data-test-configs).
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 40]
```diff
@@ -0,0 +1,306 @@
...
+
+Unit tests are essential for validating complex SQL logic and transformations in dbt models. These tests are especially valuable when working with intricate SQL expressions. They help catch errors before deploying changes, ensuring the logic behaves as expected, particularly in critical models or when handling edge cases.
+
+For more details on formatting unit tests, refer to the [official dbt documentation](https://docs.getdbt.com/reference/resource-properties/unit-tests).
+
+## Guidelines for CHT Pipeline tests
+To ensure data integrity and the reliability of the dbt models in the CHT Pipeline, it is essential to follow these testing guidelines:
+
+- **Basic generic tests** for all models:
+Every model should have generic tests to enforce critical constraints and relationships. Use dbt core build in generic tests.
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:39]:
```suggestion
Every model should have generic tests to enforce critical constraints and relationships. Use the generic tests provided in dbt core.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 41]
```diff
@@ -0,0 +1,306 @@
...
+uuid,saved_timestamp,reported,parent_uuid,name,contact_type,phone,phone2,active,notes,contact_id,muted
+c1,2024-08-01 00:00:00,2024-07-31 08:00:00+00,p1,John Doe,person,12345,54321,true,Note 1,C-123,false
+c2,2024-08-01 00:00:00,2024-07-31 08:00:00+00,p2,Jane Doe,clinic,67890,09876,true,Note 2,C-456,true
+c3,2024-08-02 00:00:00,2024-07-31 08:00:00+00,p3,Mike Smith,person,11223,33211,false,Note 3,C-789,false
+c4,2024-08-02 00:00:00,2024-07-31 08:00:00+00,p4,Sara Smith,district_hospital,44556,65544,true,Note 4,C-101,true
+```
+
+## Running tests with Docker
+
+dbt tests are run with Docker, to isolate dependencies and configurations, making the testing process reproducible and easier to manage. The scripts and configuration files below work together to automate the process of setting up a testing environment, and running dbt models and tests:
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:235]:
```suggestion
Dbt tests are run with Docker, to isolate dependencies and configurations, making the testing process reproducible and easier to manage. The scripts and configuration files below work together to automate the process of setting up a testing environment, and running dbt models and tests:
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 42]
```diff
@@ -0,0 +1,306 @@
...
+
+### Run the tests
+
+1. Navigate to `test` folder.
+2. Run the test script:
+
+```sh
+./run_dbt_tests.sh
+```
+3. After running the following shell script, you should see the Docker containers start, the tests run and pass, and finally, the Docker containers are stopped and removed. This ensures that the testing environment is clean and ready for future runs.
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:272]:
```suggestion
3. After running the shell script, you should see the Docker containers start, the tests run and pass, and finally, the Docker containers are stopped and removed. This ensures that the testing environment is clean and ready for future runs.
```

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 43]
```diff
@@ -0,0 +1,306 @@
...
+  - The results align with business expectations and requirements.
+  - Custom data tests are particularly important for aggregation models, as errors in these models can lead to significant discrepancies in reports and analyses.
+
+- **Unit tests** for complex logic:
+Unit tests are not strictly required but are highly recommended, especially for models with complex transformation logic. Examples of when to use unit tests include:
+  - Complex SQL Logic: Such as regex, date math, window functions, or extensive `CASE WHEN` statements.
+  - Custom calculations: When creating functions or applying unique data processing logic.
+  - Edge cases: To handle scenarios that are not typically found in actual data but may arise unexpectedly.
+
+## Writing CHT Pipeline tests
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:53]: The first three lines here feel repetitive because this is already mentioned in the guidelines section above. I recommend removing them altogether.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 44]
```diff
@@ -0,0 +1,306 @@
...
+
+- All models should have **basic generic tests**, specifically enforcing `unique`, `not_null`, and `relationships` where relevant.
+
+
+- **Custom data tests** for aggregation models should be added to ensure the data is as expected. Only specific models need these tests.
+
+
+- Add **unit tests** for complex logic, such as data math or window functions. These are not required but are good to have, especially if the model is covered by data tests.
+
+
```
Comment (njuguna-n) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:64]:
```suggestion
```

**Review by njuguna-n — CHANGES_REQUESTED** [review]
This is looking very good. I have made a few recommended changes.

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 45]
```diff
@@ -0,0 +1,306 @@
...
+✔ Container test-dbt-1
+Removed
+✔ Container test-postgres-1
+Removed
+✔ Network test_default
+Removed
+DBT tests passed
+```
+{{% alert title="Note" %}}
+This snippet shows the key steps of the process. The full output will contain more details, but the above highlights the main actions performed by the script.
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:305]: 👌

File: content/en/apps/guides/data/analytics/testing-dbt-models.md [hunk 46]
```diff
@@ -0,0 +1,306 @@
...
+
+1. Navigate to `test` folder.
+2. Run the test script:
+
+```sh
+./run_dbt_tests.sh
+```
+3. After running the following shell script, you should see the Docker containers start, the tests run and pass, and finally, the Docker containers are stopped and removed. This ensures that the testing environment is clean and ready for future runs.
+
+Here's a partial example of what the output should look like:
```
Comment (andrablaj) [inline content/en/apps/guides/data/analytics/testing-dbt-models.md:274]:
```suggestion
The example below is a snippet of what the output could look like:
```

**Review by andrablaj — APPROVED** [review]
I left a few nitpick comments, nothing blocking.

The page is more readable with the code blocks, thanks for adding them!

**Review by njuguna-n — APPROVED** [review]
Great job!
